### PR TITLE
Fix CohereASR training-loss double-shift (same as Moonshine fix #46784)

### LINKED ISSUE #46894: CohereASR training-loss double-shift bug (same pattern as Moonshine #46784)
### System Info

- `transformers` version: 5.13.0.dev0 (main branch at commit dc5a497c18)
- - Platform: Windows
- - - Python version: 3.x
- - - - Using GPU: N/A (code-level bug, not runtime)

### Who can help?

@eustlb @ebezzam @vasqu

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

aIn `src/transformers/models/cohere_asr/modular_cohere_asr.py`, `CohereAsrForConditionalGeneration.forward()` has a double-shift bug:

1. Labels are shifted right via `shift_tokens_right()` (line 483) to create `decoder_input_ids`
2. 2. Then `self.loss_function()` is called (line 503), which maps to `ForCausalLMLoss`
3. 3. `ForCausalLMLoss` shifts labels again internally (`labels[..., 1:]` in `loss/loss_utils.py` line 63-64)
This means the model trains against `labels[..., 1:]` instead of `labels`.

This is the **exact same bug** that was fixed in Moonshine in PR #46784 (commit d8c235494e). The Moonshine fix replaced `self.loss_function()` with `CrossEntropyLoss()`, but `CohereAsr` — which inherits from Moonshine and overrides `forward` — was not updated.

```python
# Buggy code (line 503 in modular_cohere_asr.py):
loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size)

# Should be:
loss_fct = CrossEntropyLoss()
loss = loss_fct(logits.reshape(-1, self.config.vocab_size), labels.reshape(-1))
```

### Expected behavior

Training loss should be computed directly against `labels` without any additional shifting, since `shift_tokens_right()` already handles the decoder input alignment. This matches the behavior of Whisper, Bart, and the now-fixed Moonshine model. The fix should use `CrossEntropyLoss()` instead of `self.loss_function()`, as done in Moonshine PR #46784.

### PR DESCRIPTION
**Pull request by sharmax-vikas**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=46895)

# What does this PR do?

`CohereAsrForConditionalGeneration.forward()` has a training-loss double-shift bug — the same pattern fixed in Moonshine via #46784.

**The problem:**
1. Labels are shifted right via `shift_tokens_right()` to create `decoder_input_ids`
2. Then `self.loss_function()` is called, which maps to `ForCausalLMLoss`
3. `ForCausalLMLoss` shifts labels again internally (`labels[..., 1:]`)

This means the model trains against `labels[..., 1:]` instead of `labels`.

**The fix:** Replace `self.loss_function()` with plain `CrossEntropyLoss()`, matching Whisper, Bart, and the now-fixed Moonshine model.

Fixes #46894

## Code Agent Policy

- [x] I confirm that this is not a pure code agent PR.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case. — #46894
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [ ] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

@eustlb @ebezzam @vasqu

### CODE REVIEW

File: src/transformers/models/cohere_asr/modular_cohere_asr.py [hunk 1]
```diff
@@ -500,7 +501,8 @@ def forward(
 
         loss = None
         if labels is not None:
-            loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size)
+            loss_fct = CrossEntropyLoss()
+            loss = loss_fct(logits.reshape(-1, self.config.vocab_size), labels.reshape(-1))
```
Comment (vasqu) [inline src/transformers/models/cohere_asr/modular_cohere_asr.py]: Not sure if I agree with this whole approach tbh. 

Trying to reconstruct from what I understood. 
1. We fall into the case where labels is not None but the input is None 
2. This triggers the right shift of tokens as input (with the first token being the bos token)
3. Now we apply the loss but this unnecessarily shifts the labels
4. This leads to the double shift you mentioned

However, wouldn't it make more sense to detect when we have this special case and pass it as shifted labels directly? Or am I missing something?

File: src/transformers/models/cohere_asr/modular_cohere_asr.py [hunk 2]
```diff
@@ -500,7 +500,9 @@ def forward(
 
         loss = None
         if labels is not None:
-            loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size)
+            loss = self.loss_function(
```
Comment (vasqu) [inline src/transformers/models/cohere_asr/modular_cohere_asr.py:504]: I meant it more like when we enter the case not in general, so something along

```py
        do_shift_labels = False
        if labels is not None:
            if decoder_input_ids is None and decoder_inputs_embeds is None:
                decoder_input_ids = shift_tokens_right(
                    labels, self.config.pad_token_id, self.config.decoder_start_token_id
                )
                do_shift_labels = True
        ...
        if labels is not None:
            shift_labels = kwargs.pop("shift_labels", None)
            if do_shift_labels and shift_labels is None:
                 shift_labels = labels
            loss = self.loss_function(logits=logits, labels=labels, shift_labels=shift_labels, vocab_size=self.config.vocab_size, **kwargs)
```

**Review by vasqu — COMMENTED** [review]
We also need a test! cc @ebezzam @eustlb since it seems to affect multiple models (audio?)

File: src/transformers/models/cohere_asr/modular_cohere_asr.py (hunk 2 above)
Comment (sharmax-vikas) [reply src/transformers/models/cohere_asr/modular_cohere_asr.py:504]: Thanks for the suggestion! Updated the fix to use `do_shift_labels` flag - now `shift_labels` is only passed when `shift_tokens_right()` was actually called. Also added a test (`test_training_loss_no_double_shift`). Please take a look!

**Comment by sharmax-vikas** [comment]
@vasqu

File: src/transformers/models/cohere_asr/modular_cohere_asr.py [hunk 3]
```diff
@@ -500,7 +502,16 @@ def forward(
 
         loss = None
         if labels is not None:
-            loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size)
+            shift_labels = kwargs.pop("shift_labels", None)
+            if do_shift_labels and shift_labels is None:
+                shift_labels = labels
```
Comment (SunMarc) [inline src/transformers/models/cohere_asr/modular_cohere_asr.py]: I feel like the labels should never be shifted no and only `decoder_input_ids` should be. So maybe we can even simplify to this + remove `do_shift_labels`. 

```suggestion
            shift_labels = kwargs.pop("shift_labels", labels)
```

**Review by SunMarc — COMMENTED** [review]
Left a comment

**Review by eustlb — COMMENTED** [review]
#47090 is the correct fix IMO, see https://github.com/huggingface/transformers/pull/47090#pullrequestreview-4637793592 for explaination
cc @vasqu @SunMarc if you agree

**Review by SunMarc — APPROVED** [review]
LGTM !
